File: apps/functions/src/jrdb_scraper/fetchDailyData.ts
import { logger } from 'firebase-functions'
import { downloadJRDBFile } from './downloader'
import { convertLzhToNpz } from './converter'
import { generateJRDBDataFileUrl } from './raceKeyGenerator'
import { getFirestore } from 'firebase-admin/firestore'
import { JRDBDataType } from './entities/jrdb'
import * as path from 'path'
import * as fs from 'fs'
import { createTempDir, uploadFileToStorage } from '../utils/storageUploader'
import { createDateFromYMD, formatDateISO, formatDateJRDB } from './utils/dateFormatter'
import { createStorageMetadata, createFirestoreMetadata } from './utils/metadata'
import { generateRecordKey } from './utils/recordKey'


/**
 * 単一のデータタイプのデータを取得してFirestoreに保存（内部関数）
 */
async function fetchSingleDataType(
  year: number,
  month: number,
  day: number,
  dataType: JRDBDataType,
  date: string
): Promise<{
  date: string
  dataType: string
  success: boolean
  recordCount: number
  lzhStoragePath?: string
  npzStoragePath?: string
  jsonStoragePath?: string
  fileName?: string
  error?: string
}> {
  logger.info('日単位データ取得開始', { date, dataType })

  const db = getFirestore()
  const tempDir = createTempDir()

  try {
    const actualDataType = dataType.toString()
    const dateObj = createDateFromYMD(year, month, day)
    const dateStr = formatDateJRDB(year, month, day)
    const sourceUrl = generateJRDBDataFileUrl(actualDataType, dateObj)
    const lzhBuffer = await downloadJRDBFile(sourceUrl)
    const fileName = `${actualDataType}${dateStr}`
    const npzFilePath = path.join(tempDir, `${fileName}.npz`)
    const { records } = await convertLzhToNpz(lzhBuffer, actualDataType, year, npzFilePath)

    if (records.length === 0) throw new Error('パースされたレコードが0件です')

    // 3. LZHファイル、NPZファイル、JSONファイルをStorageに保存
    
    // 3-1. LZHファイルをStorageに保存
    const lzhFilePath = path.join(tempDir, `${fileName}.lzh`)
    fs.writeFileSync(lzhFilePath, lzhBuffer)
    const lzhStoragePath = `jrdb/lzh/${fileName}.lzh`
    const lzhMetadata = createStorageMetadata(fileName, sourceUrl, actualDataType, date)
    let lzhStoragePathResult: string | undefined = lzhStoragePath
    try {
      await uploadFileToStorage(lzhFilePath, lzhStoragePath, lzhMetadata)
      logger.info('LZHファイルをStorageに保存しました', { lzhStoragePath, fileName, sourceUrl })
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      logger.error('LZHファイルのStorageへの保存に失敗しました', { fileName, error: errorMessage })
      // LZHファイルの保存失敗はエラーとしない（NPZファイルがあれば問題ない）
      lzhStoragePathResult = undefined
    }

    // 3-2. NPZファイルをStorageにアップロード
    if (!fs.existsSync(npzFilePath)) throw new Error(`NPZファイルが存在しません: ${npzFilePath}`)
    
    const npzStoragePath = `jrdb/npz/${fileName}.npz`
    const npzMetadata = createStorageMetadata(fileName, sourceUrl, actualDataType, date)
    await uploadFileToStorage(npzFilePath, npzStoragePath, npzMetadata)
    
    logger.info('NPZファイルをStorageに保存しました', { storagePath: npzStoragePath, fileName, sourceUrl })

    // 3-3. JSONファイルをStorageにアップロード
    const { convertToJson } = await import('./converter')
    const jsonFilePath = path.join(tempDir, `${fileName}.json`)
    await convertToJson(records, jsonFilePath)
    
    const jsonStoragePath = `jrdb/json/${fileName}.json`
    const jsonMetadata = createStorageMetadata(fileName, sourceUrl, actualDataType, date)
    await uploadFileToStorage(jsonFilePath, jsonStoragePath, jsonMetadata)
    
    logger.info('JSONファイルをStorageに保存しました', { storagePath: jsonStoragePath, fileName, sourceUrl })

    // 4. Firestoreにメタデータとパースしたデータを保存
    // すべてのレコードをコレクションとして統一して保存する
    const dailyDocRef = db.collection('jrdb_daily_data').doc(date)
    const dataTypeRef = dailyDocRef.collection(actualDataType.toString()).doc('metadata')

    const metadataToSave = createFirestoreMetadata(
      actualDataType,
      date,
      year,
      month,
      day,
      lzhStoragePathResult,
      npzStoragePath,
      jsonStoragePath,
      fileName,
      records.length
    )
    
    if (actualDataType !== dataType.toString()) metadataToSave.originalDataType = dataType.toString()
    
    
    await dataTypeRef.set(metadataToSave, { merge: true })

    // 各レコードを個別のドキュメントとしてサブコレクションに保存
    const recordsCollection = dailyDocRef.collection(actualDataType.toString()).doc('metadata').collection('records')
    let batch = db.batch()
    let batchCount = 0
    const maxBatchSize = 500

    for (let i = 0; i < records.length; i++) {
      const record = records[i]
      const recordKey = generateRecordKey(record, i)
      const recordRef = recordsCollection.doc(recordKey)
      batch.set(recordRef, {
        ...record,
        index: i
      })
      batchCount++

      if (batchCount >= maxBatchSize) {
        await batch.commit()
        batch = db.batch() // 新しいbatchを作成
        batchCount = 0
      }
    }

    if (batchCount > 0) await batch.commit()
    

    logger.info('レコードをコレクションとして保存しました', {
      dataType: actualDataType,
      date,
      totalRecords: records.length
    })

    // 5. 日単位のメタデータを保存
    await dailyDocRef.set({
      date,
      year,
      month,
      day,
      fetchedAt: new Date()
    }, { merge: true })

    return {
      date,
      dataType: actualDataType.toString(),
      success: true,
      recordCount: records.length,
      lzhStoragePath: lzhStoragePathResult,
      npzStoragePath: npzStoragePath,
      jsonStoragePath: jsonStoragePath,
      fileName
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error)
    logger.error('日単位データ取得エラー', { date, dataType, error: errorMessage })

    return {
      date,
      dataType: dataType.toString(),
      success: false,
      recordCount: 0,
      error: errorMessage
    }
  } finally {
    try {
      fs.rmSync(tempDir, { recursive: true, force: true })
    } catch {
      // クリーンアップエラーは無視
    }
  }
}

/**
 * 日単位で指定されたデータタイプのデータを取得してFirestoreに保存
 * @param year - 年
 * @param month - 月
 * @param day - 日
 * @param dataTypes - データタイプの配列
 */
export async function fetchDailyData(
  year: number,
  month: number,
  day: number,
  dataTypes: JRDBDataType[]
): Promise<Array<{
  date: string
  dataType: string
  success: boolean
  recordCount: number
  lzhStoragePath?: string
  npzStoragePath?: string
  jsonStoragePath?: string
  fileName?: string
  error?: string
  }>> {
  const date = formatDateISO(year, month, day)

  logger.info('日単位データ取得開始（複数データタイプ）', { date, dataTypes: dataTypes.map(dt => dt.toString()) })

  const results: Array<{
    date: string
    dataType: string
    success: boolean
    recordCount: number
    lzhStoragePath?: string
    npzStoragePath?: string
    jsonStoragePath?: string
    fileName?: string
    error?: string
  }> = []

  for (const dataType of dataTypes) 
    try {
      const result = await fetchSingleDataType(year, month, day, dataType, date)
      results.push(result)
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      logger.error('データタイプ取得エラー', { date, dataType, error: errorMessage })
      results.push({
        date,
        dataType: dataType.toString(),
        success: false,
        recordCount: 0,
        error: errorMessage
      })
    }
  

  logger.info('日単位データ取得完了', { date, totalDataTypes: dataTypes.length, successCount: results.filter(r => r.success).length })

  return results
}

File: apps/prediction/src/jrdb_scraper/fetch_daily_data.py
"""日単位データ取得
指定日付のデータタイプを取得してParquet保存
"""

import logging
from datetime import date
from pathlib import Path
from typing import Dict, List, Optional, Union

import pandas as pd

from .converter import convert_lzh_to_parquet
from .downloader import download_jrdb_file
from .entities.jrdb import JRDBDataType
from .race_key_generator import generate_jrdb_data_file_url
from .utils.date_formatter import create_date_from_ymd, format_date_iso, format_date_jrdb

logger = logging.getLogger(__name__)


def fetch_single_data_type(
    year: int,
    month: int,
    day: int,
    dataType: JRDBDataType,
    outputDir: Union[str, Path],
    dateStr: str,
    bac_df: Optional[pd.DataFrame] = None
) -> Dict[str, Union[str, int, bool, Optional[str]]]:
    """単一のデータタイプのデータを取得してParquetに保存（内部関数）
    
    Args:
        year: 年
        month: 月
        day: 日
        dataType: データタイプ
        outputDir: 出力ディレクトリ
        dateStr: 日付文字列（ISO形式）
        bac_df: BACデータ（KYI等の年月日がないデータタイプのrace_key生成に使用、オプション）
    
    Returns:
        結果辞書
    """
    logger.info('日単位データ取得開始', extra={'date': dateStr, 'dataType': dataType.value})
    
    outputPath = Path(outputDir)
    outputPath.mkdir(parents=True, exist_ok=True)
    
    try:
        actualDataType = dataType.value
        dateObj = create_date_from_ymd(year, month, day)
        dateStrJRDB = format_date_jrdb(year, month, day)
        sourceUrl = generate_jrdb_data_file_url(actualDataType, dateObj)
        lzhBuffer = download_jrdb_file(sourceUrl)
        fileName = f'{actualDataType}{dateStrJRDB}'
        
        # LZHファイルも保存（PredictionExecutorで使用するため）
        lzhFilePath = outputPath / f'{fileName}.lzh'
        with open(lzhFilePath, 'wb') as f:
            f.write(lzhBuffer)
        
        # Parquetファイルも保存
        parquetFilePath = outputPath / f'{fileName}.parquet'
        _, records = convert_lzh_to_parquet(lzhBuffer, actualDataType, year, parquetFilePath, bac_df=bac_df)
        
        if len(records) == 0:
            raise ValueError('パースされたレコードが0件です')
        
        logger.info('日単位データ取得完了', extra={
            'dataType': actualDataType,
            'date': dateStr,
            'recordCount': len(records),
            'lzhPath': str(lzhFilePath),
            'parquetPath': str(parquetFilePath)
        })
        
        return {
            'date': dateStr,
            'dataType': actualDataType,
            'success': True,
            'recordCount': len(records),
            'outputPath': str(parquetFilePath),
            'lzhPath': str(lzhFilePath),
            'fileName': fileName
        }
    except Exception as error:
        error_type = type(error).__name__
        errorMessage = f"{error_type}: {error}"
        # extra はフォーマッタによって表示されないことがあるため、メッセージ本体に含める
        # スタックトレースも必ず出す（原因調査のため）
        logger.exception(f"日単位データ取得エラー date={dateStr} dataType={dataType.value}")
        
        return {
            'date': dateStr,
            'dataType': dataType.value,
            'success': False,
            'recordCount': 0,
            'error': errorMessage
        }


def fetch_daily_data(
    year: int,
    month: int,
    day: int,
    dataTypes: List[JRDBDataType],
    outputDir: Union[str, Path] = 'cache/jrdb/parquet'
) -> List[Dict[str, Union[str, int, bool, Optional[str]]]]:
    """日単位で指定されたデータタイプのデータを取得してParquetに保存
    
    Args:
        year: 年
        month: 月
        day: 日
        dataTypes: データタイプの配列
        outputDir: 出力ディレクトリ（デフォルト: 'cache/jrdb/parquet'）
    
    Returns:
        結果のリスト
    """
    dateStr = format_date_iso(year, month, day)
    
    logger.info('日単位データ取得開始（複数データタイプ）', extra={
        'date': dateStr,
        'dataTypes': [dt.value for dt in dataTypes]
    })
    
    # KYI等の年月日がないデータタイプのrace_key生成用に、BACデータを事前に読み込む
    bac_df: Optional[pd.DataFrame] = None
    bac_result: Optional[Dict[str, Union[str, int, bool, Optional[str]]]] = None
    
    # BACが含まれている場合は、先に処理する
    if JRDBDataType.BAC in dataTypes:
        try:
            bac_result = fetch_single_data_type(year, month, day, JRDBDataType.BAC, outputDir, dateStr)
            if bac_result.get('success', False) and bac_result.get('outputPath'):
                bac_parquet_path = Path(bac_result['outputPath'])
                if bac_parquet_path.exists():
                    bac_df = pd.read_parquet(bac_parquet_path)
                    logger.info('BACデータを読み込みました（KYI等のrace_key生成用）', extra={
                        'date': dateStr,
                        'recordCount': len(bac_df) if bac_df is not None else 0
                    })
        except Exception as e:
            logger.warning('BACデータの読み込みに失敗しました（KYI等のrace_key生成に影響する可能性があります）', extra={
                'date': dateStr,
                'error': str(e)
            })
    
    results: List[Dict[str, Union[str, int, bool, Optional[str]]]] = []
    
    # BACの結果を先に追加（既に取得済みの場合）
    if bac_result is not None:
        results.append(bac_result)
    
    for dataType in dataTypes:
        # BACは既に処理済みなのでスキップ
        if dataType == JRDBDataType.BAC:
            continue
        
        try:
            # KYI等の年月日がないデータタイプの場合は、BACデータを渡す
            result = fetch_single_data_type(
                year, month, day, dataType, outputDir, dateStr,
                bac_df=bac_df
            )
            results.append(result)
        except Exception as error:
            error_type = type(error).__name__
            errorMessage = f"{error_type}: {error}"
            logger.exception(f"データタイプ取得エラー date={dateStr} dataType={dataType.value}")
            results.append({
                'date': dateStr,
                'dataType': dataType.value,
                'success': False,
                'recordCount': 0,
                'error': errorMessage
            })
    
    successCount = sum(1 for r in results if r.get('success', False))
    logger.info('日単位データ取得完了', extra={
        'date': dateStr,
        'totalDataTypes': len(dataTypes),
        'successCount': successCount
    })
    
    return results

